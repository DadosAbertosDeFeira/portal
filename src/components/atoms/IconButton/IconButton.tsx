import { Button, type ButtonProps } from "atoms/Button";
import { Image } from "atoms/Image";
import { Link, type LinkProps } from "atoms/Link";
import { Text } from "atoms/Text";
import type { StaticImageData } from "next/image";
import type { ForwardedRef } from "react";
import { forwardRef } from "react";

type IconButtonAsLink = Omit<LinkProps, "as"> & {
  as: "link";
};

type IconButtonAsButton = Omit<ButtonProps, "as"> & {
  as?: "button";
};

export type IconButtonProps = (IconButtonAsLink | IconButtonAsButton) & {
  icon: string | StaticImageData;
  label: string;
};

type HTMLButtonOrLinkElement = HTMLButtonElement | HTMLLinkElement;

export const IconButton = forwardRef<HTMLButtonOrLinkElement, IconButtonProps>(
  function IconButton(props: IconButtonProps, ref) {
    const { icon, label } = props;

    const children = (
      <>
        <Image src={icon} alt={label} />
        <Text className="text-sm text-primary-dark">{label}</Text>
      </>
    );

    if (props.as === "link") {
      return (
        <Link {...props} ref={ref as ForwardedRef<HTMLAnchorElement>}>
          {children}
        </Link>
      );
    }

    return (
      <Button {...props} ref={ref as ForwardedRef<HTMLButtonElement>}>
        {children}
      </Button>
    );
  }
);
