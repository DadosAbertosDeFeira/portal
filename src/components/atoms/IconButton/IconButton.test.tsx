import { faker } from "@faker-js/faker";
import { render, screen } from "@testing-library/react";
import type { ImageProps } from "atoms/Image";

import { IconButton, type IconButtonProps } from "./IconButton";

jest.mock("atoms/Image", () => ({
  Image: ({ src, ...props }: ImageProps) => (
    <img src={src as string} {...props} />
  ),
}));

describe("IconButton", () => {
  const makeSut = ({
    label = "test",
    icon = faker.image.url(),
    ...props
  }: Partial<IconButtonProps>) => {
    return render(<IconButton icon={icon} label={label} {...(props as {})} />);
  };

  it("renders correctly", () => {
    const { asFragment } = makeSut({});

    expect(asFragment).toMatchSnapshot();
  });

  it("should render button correctly", () => {
    makeSut({ as: "button" });

    const button = screen.queryByRole("button");

    expect(button).toBeInTheDocument();
  });

  it("should render link correctly", () => {
    makeSut({ as: "link", href: "https://link.com" });

    const link = screen.queryByRole("link");

    expect(link).toBeInTheDocument();
  });
});
