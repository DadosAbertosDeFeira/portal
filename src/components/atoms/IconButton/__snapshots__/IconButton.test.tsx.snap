// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton renders correctly 1`] = `[Function]`;
